File: routes/writingTags.js
const express = require('express');
const db = require('../db');
const { ok, fail } = require('../middleware/responses');
const { parseBooleanValue } = require('../lib/boolean');

const router = express.Router();

function mergeBool(bodyVal, key, prev) {
  if (bodyVal !== undefined) return parseBooleanValue(bodyVal);
  if (prev) return prev[key];
  return false;
}

function mergeNotes(bodyVal, prev) {
  if (bodyVal !== undefined) return bodyVal;
  if (prev) return prev.notes;
  return null;
}

router.post('/', async (req, res) => {
  const {
    entry_id,
    claim_present,
    evidence_cited,
    explanation_present,
    source_named,
    response_incomplete,
    ai_flag,
    notes,
  } = req.body;

  const eid = Number(entry_id);
  if (!Number.isInteger(eid)) return fail(res, 'entry_id is required', 400);

  const entryCheck = await db.query('SELECT 1 FROM entries WHERE id = $1', [eid]);
  if (!entryCheck.rowCount) return fail(res, 'Entry not found', 404);

  const prevResult = await db.query(
    `SELECT claim_present, evidence_cited, explanation_present, source_named,
            response_incomplete, ai_flag, notes
     FROM writing_tags WHERE entry_id = $1`,
    [eid]
  );
  const prev = prevResult.rows[0];

  const merged = {
    claim_present: mergeBool(claim_present, 'claim_present', prev),
    evidence_cited: mergeBool(evidence_cited, 'evidence_cited', prev),
    explanation_present: mergeBool(explanation_present, 'explanation_present', prev),
    source_named: mergeBool(source_named, 'source_named', prev),
    response_incomplete: mergeBool(response_incomplete, 'response_incomplete', prev),
    ai_flag: mergeBool(ai_flag, 'ai_flag', prev),
    notes: mergeNotes(notes, prev),
  };

  const { rows } = await db.query(
    `INSERT INTO writing_tags (
       entry_id, claim_present, evidence_cited, explanation_present, source_named,
       response_incomplete, ai_flag, notes
     ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
     ON CONFLICT (entry_id) DO UPDATE SET
       claim_present = EXCLUDED.claim_present,
       evidence_cited = EXCLUDED.evidence_cited,
       explanation_present = EXCLUDED.explanation_present,
       source_named = EXCLUDED.source_named,
       response_incomplete = EXCLUDED.response_incomplete,
       ai_flag = EXCLUDED.ai_flag,
       notes = EXCLUDED.notes
     RETURNING id, entry_id, claim_present, evidence_cited, explanation_present, source_named,
               response_incomplete, ai_flag, notes, created_at`,
    [
      eid,
      merged.claim_present,
      merged.evidence_cited,
      merged.explanation_present,
      merged.source_named,
      merged.response_incomplete,
      merged.ai_flag,
      merged.notes,
    ]
  );

  const status = prev ? 200 : 201;
  return ok(res, rows[0], status);
});

module.exports = router;

File: desktop/src-tauri/src/routes/writing_tags.rs
//! Port of routes/writingTags.js — upsert merging incoming fields over the
//! previous row, 201 on first insert / 200 on update.

use axum::extract::State;
use axum::http::StatusCode;
use axum::response::Response;
use rusqlite::types::Value as Sv;
use serde_json::Value;

use crate::db::{blocking, parse_boolean_value, query_json, query_json_dyn, Pool};
use crate::error::{ok_with_status, ApiError};

const TAG_KEYS: [&str; 6] = [
    "claim_present",
    "evidence_cited",
    "explanation_present",
    "source_named",
    "response_incomplete",
    "ai_flag",
];

pub async fn upsert(
    State(pool): State<Pool>,
    axum::Json(body): axum::Json<Value>,
) -> Result<Response, ApiError> {
    let Some(eid) = body["entry_id"].as_i64() else {
        return Err(ApiError::bad_request("entry_id is required"));
    };

    let (data, status) = blocking(&pool, move |conn| {
        let exists: bool = conn
            .query_row("SELECT 1 FROM entries WHERE id = ?", [eid], |_| Ok(true))
            .or_else(|e| match e {
                rusqlite::Error::QueryReturnedNoRows => Ok(false),
                other => Err(other),
            })?;
        if !exists {
            return Err(ApiError::not_found("Entry not found"));
        }

        let prev = query_json(
            conn,
            "SELECT claim_present, evidence_cited, explanation_present, source_named,
                    response_incomplete, ai_flag, notes
             FROM writing_tags WHERE entry_id = ?",
            &[&eid],
        )?
        .into_iter()
        .next();

        // Incoming field wins; otherwise previous value; otherwise false/null.
        let merge_bool = |key: &str| -> bool {
            match body.get(key) {
                Some(v) => parse_boolean_value(v),
                None => prev
                    .as_ref()
                    .and_then(|p| p[key].as_bool())
                    .unwrap_or(false),
            }
        };
        let notes: Sv = match body.get("notes") {
            Some(v) => crate::db::to_sql_value(v),
            None => prev
                .as_ref()
                .and_then(|p| p["notes"].as_str())
                .map(|s| Sv::Text(s.to_string()))
                .unwrap_or(Sv::Null),
        };

        let mut params: Vec<Sv> = vec![Sv::Integer(eid)];
        params.extend(TAG_KEYS.iter().map(|k| Sv::Integer(merge_bool(k) as i64)));
        params.push(notes);

        let rows = query_json_dyn(
            conn,
            "INSERT INTO writing_tags (
               entry_id, claim_present, evidence_cited, explanation_present, source_named,
               response_incomplete, ai_flag, notes
             ) VALUES (?, ?, ?, ?, ?, ?, ?, ?)
             ON CONFLICT (entry_id) DO UPDATE SET
               claim_present = excluded.claim_present,
               evidence_cited = excluded.evidence_cited,
               explanation_present = excluded.explanation_present,
               source_named = excluded.source_named,
               response_incomplete = excluded.response_incomplete,
               ai_flag = excluded.ai_flag,
               notes = excluded.notes
             RETURNING id, entry_id, claim_present, evidence_cited, explanation_present,
                       source_named, response_incomplete, ai_flag, notes, created_at",
            &params,
        )?;

        let status = if prev.is_some() { StatusCode::OK } else { StatusCode::CREATED };
        Ok((rows.into_iter().next().unwrap_or(Value::Null), status))
    })
    .await?;
    Ok(ok_with_status(status, data))
}
